contractsV5/staking/StakingContract.sol

```diff
@@ -0,0 +1,286 @@
+pragma solidity 0.5.11;
+
+import "../IERC20.sol";
+import "../ReentrancyGuard.sol";
+import "./IKyberStaking.sol";
+import "../DAO/IKyberDAO.sol";
+import "../EpochUtils.sol";
+
+contract StakingContract is IKyberStaking, EpochUtils, ReentrancyGuard {
+
+    // amount KNC staked of an address for each epoch
+    mapping(uint => mapping(address => uint)) internal stakes;
+    // the latest KNC staked of an address
+    mapping(address => uint) internal latestStakeBalance;
+
+    // amount of KNC that other people has delegated to an address at each epoch
+    mapping(uint => mapping(address => uint)) internal delegatedStakes;
+    // similar to latest stake balance but for delegated stakes
+    mapping(address => uint) internal latestDelegatedStakes;
+
+    // delegated address of an address at each epoch
+    mapping(uint => mapping(address => address)) internal delegatedAddress;
+    // latest delegated address of an address
+    mapping(address => address) internal latestDelegatedAddress;
+
+    // bool for control if we have init data for an epoch + an address
+    mapping(uint => mapping(address => bool)) internal hasInited;
+
+    IERC20 public KNC_TOKEN;
+    IKyberDAO public DAO;
+    address public admin;
+
+    constructor(address _kncToken, uint _epochPeriod, uint _startBlock, address _admin) public {
+        require(_epochPeriod > 0, "constructor: epoch duration must be positive");
+        require(_startBlock >= block.number, "constructor: start block should not be in the past");
+        require(_kncToken != address(0), "constructor: KNC address is missing");
+        require(_admin != address(0), "constructor: admin address is missing");
+
+        EPOCH_PERIOD = _epochPeriod;
+        START_BLOCK = _startBlock;
+        KNC_TOKEN = IERC20(_kncToken);
+        admin = _admin;
+    }
+
+    event DAOAddressSet(address _daoAddress);
+    event AdminRemoved();
+    function updateDAOAddressAndRemoveAdmin(address _daoAddress) public {
+        require(msg.sender == admin, "updateDAOAddressAndRemoveAdmin: sender address is not admin");
+        require(_daoAddress != address(0), "updateDAOAddressAndRemoveAdmin: DAO address is missing");
+
+        DAO = IKyberDAO(_daoAddress);
+        emit DAOAddressSet(_daoAddress);
+
+        // reset admin
+        admin = address(0);
+        emit AdminRemoved();
+    }
+
+    // init data if it has not been init
+    function initDataIfNeeded(address S) internal {
+        uint N = getCurrentEpochNumber();
```

Comment:
In functions that use this init, we usually have N so recalculating N here is a waste. Please consider passing block no as a param.

Reply:
Updated, passing N as current epoch to this func, as in the initial idea this func can be public so we need to compute current epoch inside the func, however we can just put it as internal func